src/glue.rs

```diff
@@ -15,24 +15,24 @@ impl Glue {
         Self { storage }
     }
 
-    pub fn execute(&mut self, sql: &str) -> Result<Payload> {
-        let parsed = parse(sql)?;
-        let statement = translate(&parsed[0])?;
-
-        let storage = self.storage.take().unwrap();
-
-        match block_on(execute(storage, &statement)) {
-            Ok((storage, payload)) => {
-                self.storage = Some(storage);
-
-                Ok(payload)
-            }
-            Err((storage, error)) => {
-                self.storage = Some(storage);
-
-                Err(error)
+    pub fn execute(&mut self, sql: &str) -> (Vec<Result<Payload>>, Option<Result<Payload>>) {
```

Comment:
I agree that the new return types can provide more info to users.
However, for this purpose, the return value needs to be a bit different.
Current `Vec<Result<Payload>>` cannot have `Err`, so it is simply enough to use `Vec<Payload>`.
The second param, `Option<Result<Payload>>` also only can be `Ok`, so `Option<Error>` is just enough.
Therefore, to represent what you wanted to say would be this below,
```rust
// return value types
pub fn execute(..) -> (Vec<Payload>, Option<Error>)
```

However, let's be more careful about the interface change.
`execute` in `Glue` is just a simple wrapper for users to run queries easily.
We usually expect function return type as a form of `Result<OkValue, ErrorValue>`.
Then, result can be simply evaluated by try operator `?`.
In this case, try operator cannot be applied to tuple form so it would be a bit inconvenient.

How about keeping the previous return type `Result<Payload>` and let `execute` to return only the last payload?
If someone needs to get all in-between `Payload` values, then they can simply run each query separately.

Reply:
How about using fail-fast system, so when multiple queries executed when there is error preceding another queries, it will return it immediately, otherwise the last query result will be returned.
I think it is more common sense that way, such as in db client like DBeaver you will experience that kind behaviour when you executed multiple query at once

Reply:
@euiko
In the prev comment, I think I was too focused on the side story so the main message was not clear.
> How about using fail-fast system, so when multiple queries executed when there is error preceding another queries, it will return it immediately, otherwise the last query result will be returned.

Thanks! I definitely agree with this.

Reply:
I **agree** to keep previous return type `Result<Payload>` and let `execute` to return only the last payload.👍
I just thought it would be nice if there was a function for execution history log just in case for people who want to trace.